mypy/subtypes.py

```diff
@@ -209,6 +217,15 @@ def visit_partial_type(self, left: PartialType) -> bool:
         # This is indeterminate as we don't really know the complete type yet.
         raise RuntimeError
 
+    def visit_type_type(self, left: TypeType) -> bool:
+        right = self.right
+        if isinstance(right, TypeType):
+            return is_subtype(left.item, right.item)
+        if isinstance(right, CallableType):
+            return right.is_type_obj() and is_subtype(left.item, right.ret_type)
+        # XXX Others? Union, Any, TypeVar
```

Comment:
Thanks! You hit the nail right on the head with all three passages of
dubious code. Independently I realized the same and came up with nearly the
same fixes (your fix for constraints.py is better).

Now I'm stuck with one remaining mystery. Why does the reveal_type() call
produce ProUser* rather than plain ProUser? The '*' is appended by
types.py:1020 when t.erased -- what exactly does that mean and should we
expect it here?